### ISSUE TITLE
Setting the default Fido download directory to not ~/sunpy

### ISSUE THREAD

**Issue by dstansby** [sunpy/sunpy#4117]
I'm currently working on writing an external Fido client here https://github.com/heliopython/heliopy/pull/877, but ideally I would like data downloaded with clients in `heliopy` to be downloaded to `~/heliopy` instead of `~/sunpy` (or a user supplied path in the heliopy config). I don't think it's currently possible to do this.

**Comment by Cadair** [comment]
I am interpreting this as you want to be able to set the download directory per-client in Fido.

I am not sure this is going to be possible or desirable at all. The path handling is done at the `Fido` layer, and having behaviour which varies based on client I think would be confusing to the user. Currently we can document the default path, and then if a path isn't specified it all goes there. Providing the users insight into the expected default download location if it varies depending on what the search is would be hard to impossible. Fido is already pretty complex because of its design.

The only sane alternative I can think of is changing the default download dir to be something more neutral e.g. `~fido_downloads`, I am in principal happy with this but I worry about backwards compatibility with this, as it would potentially cause people to re-download stuff.

**Comment by ehsteve** [comment]
Shouldn't this be configurable with the config file [sunpyrc](https://github.com/sunpy/sunpy/blob/7c48c8ca44e11561a71e2a44c605a8d1b1a01575/sunpy/data/sunpyrc)? Then it would just be a matter of packages shipping with their own custom config. I highly recommend if we cannot do this we should enable it.

**Comment by Cadair** [comment]
The sunpyrc is for users not packages though. If heliopy changed that then all the fido downloads for sunpy would also end up in ~/heliopy`

**Comment by dstansby** [comment]
> I am interpreting this as you want to be able to set the download directory per-client in Fido.

Correct, or per package (ie. a different directory for sunpy/heliopy).

> The path handling is done at the `Fido` layer, and having behaviour which varies based on client I think would be confusing to the user.

The use case I am thinking of is a planetary scientist who wants a bunch of cassini data, and therefore doesn't care about sunpy. I think it would be confusing in this case to download data to a `sunpy` directory.

Perhaps the answer is that heliopy has it's own `Fido` instance with a different download directory set? This means if a user wants to search for sunpy and heliopy data they have to use a different search, but that might be more transparent anyway?

**Comment by Cadair** [comment]
> if a user wants to search for sunpy and heliopy data they have to use a different search

This kinda rather defeats the whole purpose of a "unified downloader" no? Maybe they do want to just download cassini data today, but maybe tomorrow they also care about the code X ray light curve to correlate with some thing in their data? :shrug: The point being the idea from my perspective is that we have one interface to the data.

> I think it would be confusing in this case to download data to a sunpy directory.

I don't disagree with this, but I do think there can only be one default dir for Fido, the same default dir you use if you download data with a client in sunpy core, dkist, or heliopy. Maybe the solution is just to make that dir independent of a package name like "sunpy"?

**Comment by ehsteve** [comment]
How about we call that directory “Fido”. Problem solved! The subdir could be the name of the package so “Fido/sunpy” if needed.

**Comment by dstansby** [comment]
Coming back to this now I'm writing some clients for `heliopy`, it would be nice to be able to specify a default download directory as an attribute of `UnifiedDownloaderFactory`, so one can create different `UnifiedDownloaderFactory`s with different default download paths.

**Comment by Cadair** [comment]
The idea of having multiple instances of `UnifiedDownloaderFactory` is something I have never considered before and sounds decidedly weird. What about just changing the default to be package agnostic?

**Comment by dstansby** [comment]
Sorry, that was a bit stream of thought this morning, and I agree that multiple `UnifiedDownloaderFactory` weird... package agnostic would be good.

Another thought I had was to allow clients to have their own sub-directories, e.g. `my_sunpy_data_dir/solar_orbiter` or similar. I do think it's quite tedious having **all** data downloaded to the same directory at the moment.

**Comment by Cadair** [comment]
The issue with being more specific than one base dir by default is you are requiring all the return tables from all the clients returning for that search to have that key, so if you wanted the default to be "~/Fido/{Source}/{file}` then you would have to mandate that all results always included a `Source` column.

**Comment by Cadair** [comment]
I have just come back to this with my dkist hat on. Not because I want to change the default base path (`~/sunpy`) but because I want to change the default path template from `{download_dir}/{file}` to `{download_dir}/{dataset_id}/{file}`.

Given that heliopy has largely migrated into sunpy, what are your current thoughts on this issue @dstansby ?

**Comment by SolarDrew** [comment]
To me it looks like the obvious way to do this would be to change it in `GenericClient`, but then it gets skipped when subclasses overwrite `fetch()`, so I assume I'm missing something. What would be the recommended way to make this change for all clients?
